---
title: "Assignment 2: Differential Gene Expression and Preliminary ORA"
author: "Luke Zhang"
date: "3/3/2020"
output:
    html_document:
        toc: true
        toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Note: The documentation makes references to the figures directory in my repo as Figure <number>. The corresponding image file is named <number>_<filename>.<format>. For example, Figure 1 is my pre-normalization MDS plot, and corresponds to the file named figures/1_pre-norm_mds.png.

## Introduction

Begin by producing the cleaned, normalized, and mapped dataset from Assignment 1:

```{r a1, results='hide', message=FALSE}
source("data_processing.R")
```

data_processing.R is a script that implements the workflow performed in Assignment 1. It downloads the supplementary data file of GEO dataset [GSE110021](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE110021).

The experiment treats an experimental group of WI-38 fibroblasts with TGFb (transforming growth factor beta) and measures gene expression in comparison to the untreated control group. Gene profiles are generated at 1-day and 20-day time points, so the experiment can be split into 4 groups. Each group has 3 replicates, so our experiment has 12 samples.

The experiment protocol has 4 columns per sample (totalling 48 columns in the downloaded data), corresponding to each lane on the sequencing machine. The script first collapses these individual lanes into a single column for the sample, resulting in an initial data frame of 12 columns (samples) and 25702 genes. Using edgeR, we filter out any genes of total CPM < 3, which is the size of our smallest group of replicates. This removes 12443 uninformative features, and we are left with 13259 genes. The remaining genes are normalized with respect to Treatment Type via edgeR TMM. Finally, the normalized genes have their given Ensemble gene IDs mapped to HGNC symbols, and any unmapped genes were removed. We also ensured that no two genes got mapped to the same symbol.

The final result is a data frame of 12 columns and 12914 rows, with unique HUGO identifiers as rownames. We will use this data frame in this assignment. Here are the first 10 rows and last 6 columns (day-20 for both treatments) of the data frame, to get an idea of what it looks like:

```{r}
knitr::kable(mapped_data[1:10, 7:12])
```

## Differential Gene Expression

Our analysis has defined noTGFb and TGFb as its groups, regardless of time. Our goal is to uncover time-independent DEGs in our samples by looking for significantly up/down-regulated genes after treatment with TGFb. Figures 1 and 2 are our previously generated MDS plots, pre- and post-normalization, and the neat clustering with respect to treatment type justifies the usage of Treatment as our principal factor. Based on this, we should be able to get good results in the upcoming analysis.

Construct our model and run the differential analysis.
```{r}
library(edgeR)

d <- DGEList(counts=mapped_data, group=samples$Treatment)
# construct model
design <- model.matrix(~ samples$Day + samples$Treatment)
knitr::kable(design)

# estimate dispersion
d <- estimateDisp(d, design)
# calculate normalization factors
d <- calcNormFactors(d)
# fit model to data
fit <- glmQLFit(d, design)
# calculate differential expression
qlf_test <- glmQLFTest(fit, coef="samples$TreatmentTGFb")

# Get the results
qlf_output <- topTags(qlf_test, sort.by="PValue", n=nrow(mapped_data))
```

Let's see what our top results look like:
```{r}
# What does this look like?
knitr::kable(topTags(qlf_test, sort.by="PValue", n=20))
```

The glmQLFTest has a built in p-value correction called FDR, which uses Benjamini-Hochberg's multiple testing method to lower false discovery rate of significantly expressed genes. We will use this parameter to threshold our corrected gene p-values. Both p-value and FDR are thresholded at < 0.05, an arbitrary threshold, but one that has been the standard in most areas of statistical testing.

```{r}
# How many genes have p < 0.05?
length(which(qlf_output$table$PValue < 0.05))

# How many genes have FDR < 0.05?
length(which(qlf_output$table$FDR < 0.05))
```

We see that 2925 genes pass the threshold p-value, but after adjustment, this number drops to 467. We conclude that these 467 genes are my significantly differentially expressed genes. Get a separate list of up and down regulated genes.

```{r}
# Upregulated genes.
upreg_genes <- qlf_output[qlf_output$table$FDR < 0.05 & qlf_output$table$logFC > 0, ]
# How many?
nrow(upreg_genes)

# Down regulated genes.
downreg_genes <- qlf_output[qlf_output$table$FDR < 0.05 & qlf_output$table$logFC < 0, ]
# How many?
nrow(downreg_genes)
```

We have 285 upregulated genes and 182 downregulated genes after treatment.

## Thresholded Over-Representation Analysis

## Interpretation

## References